### PR TITLE
feat(test-vectors): generate test-vectors

### LINKED ISSUE #302: Test vector for peg-handoff wire format

### PR DESCRIPTION
**Pull request by soju-drinker**
Refactors how test-vectors are generated and adds new test-vectors for peg-in, peg-out request and peg-handoff.

Closes #302

### CODE REVIEW

File: test-vectors/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+mod peg_handoff;
+mod peg_in;
+mod peg_out;
+mod utils;
```
Comment (netrome) [inline test-vectors/src/lib.rs:4]: fix: These modules should be public to expose the functions
```suggestion
pub mod peg_handoff;
pub mod peg_in;
pub mod peg_out;
pub mod utils;
```

**Comment by MarvinJanssen** [comment]
Can the generated test vectors be checked into the repo for easy access? For example hex-encoded in the `README.md`?

File: test-vectors/src/peg_handoff.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    generate_test_vector(input, output)
+}
+
+fn generate_peg_handoff_output(peg_wallet_address: [u8; 32], reward_cycle: u8) -> Vec<TxOut> {
+    let op_bytes = [105, 100, b'H'];
+
+    let op_return_script = Builder::new()
+        .push_opcode(opcodes::all::OP_RETURN)
+        .push_slice::<[u8; 3]>(op_bytes)
+        .push_slice::<[u8; 1]>([reward_cycle])
```
Comment (netrome) [inline test-vectors/src/peg_handoff.rs:31]: Nice with the explicit slice size annotations :star:

Comment (netrome) [reply test-vectors/src/peg_handoff.rs:31]: The reward cycle should be a u64 though.
```suggestion
        .push_slice::<[u8; 3]>(op_bytes)
        .push_slice(reward_cycle.as_be_bytes())
```

File: test-vectors/src/peg_handoff.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    let peg_wallet_address = [4; 32];
+    let reward_cycle = 67;
+
+    let output = generate_peg_handoff_output(peg_wallet_address, reward_cycle);
+
+    generate_test_vector(input, output)
+}
+
+fn generate_peg_handoff_output(peg_wallet_address: [u8; 32], reward_cycle: u8) -> Vec<TxOut> {
```
Comment (netrome) [inline test-vectors/src/peg_handoff.rs:25]:
```suggestion
fn generate_peg_handoff_output(peg_wallet_address: [u8; 32], reward_cycle: u64) -> Vec<TxOut> {
```

File: test-vectors/src/peg_handoff.rs (hunk 2 above)
Comment (netrome) [reply test-vectors/src/peg_handoff.rs:31]: The reward cycle should be a u64 though.
```suggestion
        .push_slice::<[u8; 3]>(op_bytes)
        .push_slice(reward_cycle.as_be_bytes())
```

File: test-vectors/src/peg_in.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    opcodes, script::Builder, OutPoint, Script, Sequence, Transaction, TxIn, TxOut, Witness,
+};
+
+use crate::utils::generate_test_vector;
+
+pub fn generate_peg_in_test_vector() -> Transaction {
+    let input = TxIn {
+        previous_output: OutPoint::null(),
+        script_sig: Script::empty().into(),
+        sequence: Sequence(65535),
```
Comment (netrome) [inline test-vectors/src/peg_in.rs:11]:
```suggestion
        sequence: Sequence::MAX,
```

File: test-vectors/src/peg_in.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    // Random generated wallet address
+    let stacks_address = [1u8; 20];
+    let current_peg_wallet = [4; 32];
+
+    let output = generate_peg_in_request_output(stacks_address, current_peg_wallet);
+
+    generate_test_vector(input, output)
+}
+
+fn generate_peg_in_request_output(
```
Comment (netrome) [inline test-vectors/src/peg_in.rs:26]: The peg in operation is not a request in the current design.
```suggestion
fn generate_peg_in_output(
```

File: test-vectors/src/peg_in.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    stacks_add: [u8; 20],
```
Comment (netrome) [inline test-vectors/src/peg_in.rs:27]: I don't see any point in abbreviating this name
```suggestion
    stacks_address: [u8; 20],
```

File: test-vectors/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,28 @@
+use bitcoin::{consensus::encode::serialize, Transaction, TxIn, TxOut};
```
Comment (netrome) [inline test-vectors/src/utils.rs:1]: Nit: From a code organization perspective I think this should be a private module. However, I assume the `serialize_tx` function is external so it should either be publicly re-exported or implemented somewhere else.

**Review by netrome — CHANGES_REQUESTED** [review]
Nice stuff. Just fix the u8->u64 and omit "request" from the peg in function and we're good to go.

File: test-vectors/src/utils.rs (hunk 7 above)
Comment (soju-drinker) [reply test-vectors/src/utils.rs:1]: Yeah I agree with that

**Review by jferrant — APPROVED** [review]
LGTM :D

File: test-vectors/src/peg_in.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    generate_test_vector(input, output)
+}
+
+fn generate_peg_in_output(stacks_address: [u8; 20], current_peg_wallet: [u8; 32]) -> Vec<TxOut> {
+    let op_bytes = [105, 100, b'<'];
+
+    let op_return_script = Builder::new()
+        .push_opcode(opcodes::all::OP_RETURN)
+        .push_slice::<[u8; 3]>(op_bytes)
+        .push_slice::<[u8; 20]>(stacks_address)
```
Comment (netrome) [inline test-vectors/src/peg_in.rs:32]: I forgot the version byte when we last talked about this. The version byte should be encoded in the bytes. Can you add a version byte with version 26 encoded as a single leading byte to the stacks address?

Why 26? It's the constant `C32_ADDRESS_VERSION_TESTNET_SINGLESIG` defined in `stacks-common/src/address/mod.rs`

Comment (soju-drinker) [reply test-vectors/src/peg_in.rs:32]: Sure! Just curious, so this version is going to be different on mainnet?

Comment (netrome) [reply test-vectors/src/peg_in.rs:42]: Yep, it will.

**Review by netrome — APPROVED** [review]

**Pull request merged**